### ISSUE TITLE
[Suggestion] Automatic boxed value pooling for small types (bool and bool?)

### ISSUE THREAD

**Issue by hrumhurum** [dotnet/runtime#47596]
Please take a look at the snippet:

```csharp
var obj1 = (object)true;
var obj2 = (object)true;
Console.WriteLine(ReferenceEquals(obj1, obj2));
```

Currently this code produces `false` (.NET 5.0) signifying that for a single `true` Boolean value two boxed objects were created. However in real code it would be highly beneficial to have pooled objects for small types like `bool` and `bool?`. Sample scenario:

```csharp
// There are millions of 'obj' instances.

var annotations = obj.Annotations;

bool? value = (bool?)annotations[key];
if (value == null)
{
   value = SlowCalculationFunction();

   // Cache the value for future re-use.
   annotations[key] = value;
}

// Use a calculated 'value' here.
```

What I suggest is that the following code would produce `true`:

```csharp
var obj1 = (object)true;
var obj2 = (object)true;
Console.WriteLine(ReferenceEquals(obj1, obj2));
```

This would allow to achieve a significant memory preservation for some boxed primitive value types.

**Comment by Clockwork-Muse** [comment]
... In Java, this is enabled by having the box referenced objects return a cached value when requested by the programmer. 
That is:
```java
// generates new object
Object a = (Object) some_variable;
// reuses cached object
Object b = Boolean.value(some_variable);
// the cached object
Object c = Boolean.TRUE;
```
... and pretty much all the primitives have something similar (although most only cache a small number of values, for obvious reasons).

> This would allow to achieve a significant memory preservation for some boxed primitive value types.

Setting up your own pool is generally trivial, especially as object references are generally best avoided (so in theory less need for this).
While it may save memory (barring some other optimization, like just storing some flag in the memory address), a potential issue is that now the object is not local to whatever else you're doing, which may cause its own set of issues.
It does have the potential to break something that's relying on them being different, which seems... unlikely, and possibly shortsided.

**Comment by vcsjones** [comment]
This was discussed in some depth in #7079.

Of particular note:

> Boxing is required in the C# language specification and in the CLR specification to allocate a new object instance.

**Comment by hrumhurum** [comment]
The automatic nature of such an improvement does not look so bright now due to the CLR spec that insists on `box` always returning a new instance. Thank you for pointing that out.

Regarding possible BCL solution: this sounds good.

Following a Java example, something like this would fit .NET:

```csharp
// generates new object
var a = (object)some_variable;
// reuses cached object
var b = Boolean.Boxed.Value(some_variable);
// the cached object
var c = Boolean.Boxed.True;
```

In this way a developer can make a sane choice in favor of pooled boxing depending on particular situation.

**Comment by huoyaoyuan** [comment]
This pattern is already used in libraries like WPF.
If the change can't be applied to `box` instruction, then it's benefit is quite limited.

**Comment by srxqds** [comment]
good idea

**Comment by EgorBo** [comment]
I decided to give it a try and made a prototype for JIT/VM: https://github.com/EgorBo/runtime-1/commit/0bcfcd7e31039276cdce8d96a3c645ab42a5a643 (to cache boxed `true`, `false` booleans and `0..9` integers) 

Example:
```csharp
object Test(bool x) => x;
```
Current codegen:
```asm
; Method Tests:Test(bool):System.Object:this
G_M5700_IG01:   
       56                   push     rsi
       4883EC20             sub      rsp, 32
       8BF2                 mov      esi, edx
G_M5700_IG02: 
       48B9B8075872F97F0000 mov      rcx, 0x7FF9725807B8
       E86A61515F           call     CORINFO_HELP_NEWSFAST  ;; <--- allocation
       40887008             mov      byte  ptr [rax+8], sil
G_M5700_IG03:
       4883C420             add      rsp, 32
       5E                   pop      rsi
       C3                   ret      
; Total bytes of code: 32
```
New codegen:
```asm
; Method Tests:Test(bool):System.Object:this
G_M5700_IG01: 
G_M5700_IG02:
       84D2                 test     dl, dl
       750F                 jne      SHORT G_M5700_IG04
G_M5700_IG03: 
       48B8C09294EDD5010000 mov      rax, 0x1D5ED9492C0  ;; <-- hardcoded handle for "(object)false" object
       488B00               mov      rax, gword ptr [rax]
       EB0D                 jmp      SHORT G_M5700_IG05
G_M5700_IG04: 
       48B8C89294EDD5010000 mov      rax, 0x1D5ED9492C8  ;; <-- hardcoded handle for "(object)true" object
       488B00               mov      rax, gword ptr [rax]
G_M5700_IG05: 
       C3                   ret      
; Total bytes of code: 33
```
(zero-alloc)

For `Int32` it's going to emit something like this (not implemented yet):
```csharp
if ((uint)x < 10) // can be wider range, e.g. -127..128
    return cachedObjects + x; // cached [0..9] objects
return box(x); // fallback for other values
```

If I implement it for integers today, I'll try to catch cases where it all allows to save some noticeable amount of allocations in various open-source apps. 

/cc @jkotas @benaadams (as an author of https://github.com/dotnet/runtime/issues/7079)

**Comment by jkotas** [comment]
Yep, the concerns discussed in #7079 still apply.

**Comment by nietras** [comment]
Please dont do this for any type. We mutate boxed value types constantly in our high performance processing libs and this means, boxed values must be unique and new. The BCL actually has code that does this too. If anything this should be surfaced as a lib method. 😅

This is easily fixed in C# directly and does not require JIT help. IMHO it would be better to focus on issues with that JIT that cannot easily be worked around in c# i. e. big value type issues.

**Comment by huoyaoyuan** [comment]
This should be definitely only applicable to `readonly struct`.

**Comment by nietras** [comment]
> This should be definitely only applicable to readonly struct.

@huoyaoyuan doesn't really matter if readonly or not.
```csharp 
var boxedNumber = (object)new Number(42);
ref var refNumber = ref Unsafe.Unbox<Number>(boxedNumber);
refNumber = new Number(17);
System.Console.WriteLine(boxedNumber);

readonly struct Number
{
    public Number(long value) => Value = value;
    public long Value { get; }
    public override string ToString() => Value.ToString();
}
```
This will print `17`. It's a fully supported scenario.

**Comment by EgorBo** [comment]
> > This should be definitely only applicable to readonly struct.
> 
> @huoyaoyuan doesn't really matter if readonly or not.
> 
> ```cs
> var boxedNumber = (object)new Number(42);
> ref var refNumber = ref Unsafe.Unbox<Number>(boxedNumber);
> refNumber = new Number(17);
> System.Console.WriteLine(boxedNumber);
> 
> readonly struct Number
> {
>     public Number(long value) => Value = value;
>     public long Value { get; }
>     public override string ToString() => Value.ToString();
> }
> ```
> 
> This will print `17`. It's a fully supported scenario.

That is ok, if you use Unsafe (and Unsafe.Unbox specifically) you're your own worst enemy :-) you can do exactly the same with string literals. Btw, BCL doesn't use that API anywhere.

**Comment by tannergooding** [comment]
I don't think so. The concerns @jkotas raised still apply and if we did something here it'd be a general optimization across far more than just `bool` and `bool?`
